File: frontend/src/app/api/resource/analysis-resource.ts
import {BaseResource} from './base/BaseResource';
import {AnalysisResponse} from '../dto/analysis/analysis-response';
import {ApiService} from '../services/api.service';
import {map} from 'rxjs/operators';
import {Observable} from 'rxjs/Observable';
import {WorkflowQuery} from '../dto/analysis/workflow/workflow-query';
import {Project} from '../models/Project';
import {NgxDataSet} from '../dto/ngx/grouped-series/ngx-data-set';
import {WipQuery} from '../dto/analysis/wip/wip-query';
import {WipResponse} from '../dto/analysis/wip/wip-response';
import {TimeQuery} from '../dto/analysis/time/time-query';
import {TimeResponse} from '../dto/analysis/time/time-response';
import {DeveloperRatioQuery} from '../dto/analysis/devRatio/developer-ratio-query';
import {DeveloperRatioResponse} from '../dto/analysis/devRatio/developer-ratio-response';

export class AnalysisResource extends BaseResource<AnalysisResponse> {

  constructor(api: ApiService) {
    super("Analysis", api);
  }

  getProjects (): Observable<Project[]> {
    return this.api.httpClient.get<Project[]>(this.url + "/Project", { headers: this.getHeaders(false)})
      .pipe(map(content => this.deserialize(content)));
  }

  getWorkFlow (entity: WorkflowQuery): Observable<NgxDataSet[]> {
    return this.api.httpClient.put<NgxDataSet[]>(this.url + "/WorkFlow", this.serialize(entity), { headers: this.getHeaders(false)})
      .pipe(map(content => this.deserialize(content)));
  }

  getWip (entity: WipQuery): Observable<WipResponse> {
    return this.api.httpClient.put<WipResponse>(this.url + "/Wip", this.serialize(entity), { headers: this.getHeaders(false)})
      .pipe(map(content => this.deserialize(content)));
  }

  getTime (entity: TimeQuery): Observable<TimeResponse> {
    return this.api.httpClient.put<TimeResponse>(this.url + "/Time", this.serialize(entity), { headers: this.getHeaders(false)})
      .pipe(map(content => this.deserialize(content)));
  }

  getDevRatio (entity: DeveloperRatioQuery): Observable<DeveloperRatioResponse> {
    return this.api.httpClient.put<DeveloperRatioResponse>(this.url + "/DevRatio", this.serialize(entity), { headers: this.getHeaders(false)})
      .pipe(map(content => this.deserialize(content)));
  }
}

File: backend/server/rest/src/main/java/si/fri/smrpo/kis/server/rest/resources/entities/AnalysisResource.java
package si.fri.smrpo.kis.server.rest.resources.entities;

import org.keycloak.KeycloakPrincipal;
import si.fri.smrpo.kis.core.rest.resource.base.BaseResource;
import si.fri.smrpo.kis.server.ejb.models.analysis.devRatio.DeveloperRatioQuery;
import si.fri.smrpo.kis.server.ejb.models.analysis.devRatio.DeveloperRatioResponse;
import si.fri.smrpo.kis.server.ejb.models.analysis.time.TimeQuery;
import si.fri.smrpo.kis.server.ejb.models.analysis.time.TimeResponse;
import si.fri.smrpo.kis.server.ejb.models.analysis.wip.WipQuery;
import si.fri.smrpo.kis.server.ejb.models.analysis.wip.WipResponse;
import si.fri.smrpo.kis.server.ejb.models.analysis.workflow.WorkFlowQuery;
import si.fri.smrpo.kis.server.ejb.models.analysis.workflow.WorkFlowResponse;
import si.fri.smrpo.kis.server.ejb.service.interfaces.AnalysisServiceLocal;
import si.fri.smrpo.kis.server.jpa.entities.UserAccount;
import si.fri.smrpo.kis.server.rest.resources.utils.KeycloakAuth;

import javax.ejb.EJB;
import javax.enterprise.context.RequestScoped;
import javax.ws.rs.GET;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.core.Response;

@Path("Analysis")
@RequestScoped
public class AnalysisResource extends BaseResource {

    @EJB
    private AnalysisServiceLocal service;


    private UserAccount getAuthUser() {
        return KeycloakAuth.buildAuthUser((KeycloakPrincipal) sc.getUserPrincipal());
    }

    @Path("Project")
    @GET
    public Response getProjects() throws Exception {
        return Response.status(Response.Status.OK).entity(service.getProjects(getAuthUser())).build();
    }

    @Path("WorkFlow")
    @PUT
    public Response getWorkFlowAnalysis(WorkFlowQuery query) throws Exception {
        WorkFlowResponse response = service.processWorkFlowResponse(query, getAuthUser());
        return Response.status(Response.Status.OK).entity(response.getDates()).build();
    }

    @Path("Wip")
    @PUT
    public Response getWipAnalysis(WipQuery query) throws Exception {
        WipResponse response = service.processWipResponse(query, getAuthUser());
        return Response.status(Response.Status.OK).entity(response).build();
    }

    @Path("Time")
    @PUT
    public Response getTimeAnalysis(TimeQuery query) throws Exception {
        TimeResponse response = service.processTimeResponse(query, getAuthUser());
        return Response.status(Response.Status.OK).entity(response).build();
    }

    @Path("DevRatio")
    @PUT
    public Response getDevRatio(DeveloperRatioQuery query) throws Exception {
        DeveloperRatioResponse response = service.processDeveloperRatio(query, getAuthUser());
        return Response.status(Response.Status.OK).entity(response).build();
    }

}
